File: backend/src/submission/submission.controller.ts
/* eslint-disable @typescript-eslint/ban-ts-comment */
// @ts-nocheck
import {
  Controller,
  Get,
  Param,
  ParseIntPipe,
  Post,
  Body,
  Query,
  Req,
  UseGuards,
} from '@nestjs/common';
import {
  ApiTags,
  ApiOperation,
  ApiResponse,
  ApiBearerAuth,
  ApiParam,
} from '@nestjs/swagger';
import type { Request } from 'express';
import { SubmissionService } from './submission.service';
import { ContestSubmissionService } from './contest-submission.service';
import { AuthGuard } from '../auth/auth.guard';
import { CreateSubmissionDto } from './dto/create-submission.dto';
import { RunSubmissionDto } from './dto/run-submission.dto';
import {
  GetDailyActivityQueryDto,
  FindAllSubmissionsQueryDto,
  ProblemSubmissionsQueryDto,
} from './dto/submission-query.dto';
import { Locale } from '../i18n/i18n.decorator';
import type { SupportedLocale } from '../i18n/i18n.constants';
import { ThrottleSubmission } from '../common/guards/throttle.guard';

interface AuthenticatedRequest extends Request {
  user: { id: string };
}

@ApiTags('submissions')
@Controller('submissions')
@ApiBearerAuth()
export class SubmissionController {
  constructor(private readonly submissionService: SubmissionService) {}

  @Get('status/map')
  @UseGuards(AuthGuard)
  @ApiOperation({
    summary: 'Get problem status map',
    description: 'Get submission status for all problems',
  })
  @ApiResponse({ status: 200, description: 'Map of problem ID to status' })
  async getStatusMap(@Req() req: AuthenticatedRequest) {
    const userId = req.user.id;
    const map = await this.submissionService.getProblemStatusMap(userId);
    // Convert Map to object for JSON response
    return Object.fromEntries(map);
  }

  @Get('statuses')
  @ApiOperation({
    summary: 'Get status definitions',
    description: 'Get all submission status definitions',
  })
  @ApiResponse({ status: 200, description: 'List of status definitions' })
  async getStatuses(@Locale() locale?: string) {
    return this.submissionService.getStatusDefinitions(
      locale as SupportedLocale,
    );
  }

  @Get('calendar')
  @UseGuards(AuthGuard)
  @ApiOperation({
    summary: 'Get daily activity',
    description: 'Get submission activity calendar for a year',
  })
  @ApiResponse({ status: 200, description: 'Daily submission activity' })
  async getDailyActivity(
    @Query() query: GetDailyActivityQueryDto,
    @Req() req: AuthenticatedRequest,
  ) {
    const userId = req.user.id;
    const yearInt = query.year ?? new Date().getFullYear();
    return this.submissionService.getDailyActivity(userId, yearInt);
  }

  @Get('history')
  @UseGuards(AuthGuard)
  @ApiOperation({
    summary: 'Get submission history',
    description: 'Get user submission history',
  })
  @ApiResponse({ status: 200, description: 'Submission history' })
  async getSubmissionHistory(@Req() req: AuthenticatedRequest) {
    const userId = req.user.id;
    return this.submissionService.getSubmissionHistory(userId);
  }

  @Get('learning-progress')
  @UseGuards(AuthGuard)
  @ApiOperation({
    summary: 'Get learning progress',
    description: 'Get user learning progress statistics',
  })
  @ApiResponse({ status: 200, description: 'Learning progress data' })
  async getLearningProgress(@Req() req: AuthenticatedRequest) {
    const userId = req.user.id;
    return this.submissionService.getLearningProgress(userId);
  }

  @Get(':id')
  @UseGuards(AuthGuard)
  @ApiOperation({
    summary: 'Get submission by ID',
    description: 'Retrieve a specific submission',
  })
  @ApiParam({ name: 'id', description: 'Submission ID', type: String })
  @ApiResponse({ status: 200, description: 'Submission details' })
  @ApiResponse({ status: 404, description: 'Submission not found' })
  async findOne(@Param('id') id: string, @Req() req: AuthenticatedRequest) {
    return this.submissionService.findOne(id, req.user.id);
  }

  @Get()
  @UseGuards(AuthGuard)
  @ApiOperation({
    summary: 'Get user submissions',
    description: 'Retrieve paginated submissions for the authenticated user',
  })
  @ApiResponse({ status: 200, description: 'List of submissions' })
  async findAllByUser(
    @Query() query: FindAllSubmissionsQueryDto,
    @Req() req: AuthenticatedRequest,
  ) {
    const userId = req.user.id;
    if (query.best === 'true' && query.problemId) {
      return this.submissionService.findBest(query.problemId, userId);
    }
    return this.submissionService.findAll(
      query.problemId ?? null,
      userId,
      query.skip ?? 0,
      query.take ?? 10,
    );
  }
}

@Controller('problems/:problemId/submissions')
export class ProblemSubmissionController {
  constructor(private readonly submissionService: SubmissionService) {}

  @Get()
  @UseGuards(AuthGuard)
  async findAll(
    @Param('problemId', ParseIntPipe) problemId: number,
    @Query() query: ProblemSubmissionsQueryDto,
    @Req() req: AuthenticatedRequest,
  ) {
    const userId = req.user.id;
    return this.submissionService.findAll(
      problemId,
      userId,
      query.skip ?? 0,
      query.take ?? 10,
    );
  }

  @Get('best')
  @UseGuards(AuthGuard)
  async findBest(
    @Param('problemId', ParseIntPipe) problemId: number,
    @Req() req: AuthenticatedRequest,
  ) {
    const userId = req.user.id;
    return this.submissionService.findBest(problemId, userId);
  }

  @Post('run')
  @UseGuards(AuthGuard)
  @ThrottleSubmission()
  async run(
    @Param('problemId', ParseIntPipe) problemId: number,
    @Body() dto: RunSubmissionDto,
    @Req() req: AuthenticatedRequest,
  ) {
    return this.submissionService.run(problemId, dto, req.user.id);
  }

  @Post()
  @UseGuards(AuthGuard)
  @ThrottleSubmission()
  async create(
    @Param('problemId', ParseIntPipe) problemId: number,
    @Body() dto: CreateSubmissionDto,
    @Req() req: AuthenticatedRequest,
  ) {
    const userId = req.user.id;
    return this.submissionService.create(userId, problemId, dto);
  }
}

/**
 * Contest-specific submission endpoints
 */
@Controller('contest/:contestId/problems/:problemId/submissions')
export class ContestSubmissionController {
  constructor(
    private readonly contestSubmissionService: ContestSubmissionService,
  ) {}

  @Get()
  @UseGuards(AuthGuard)
  async getContestSubmissions(
    @Param('contestId') contestId: string,
    @Param('problemId', ParseIntPipe) problemId: number,
    @Req() req: AuthenticatedRequest,
  ) {
    const userId = req.user.id;
    return this.contestSubmissionService.getContestSubmissions(
      contestId,
      userId,
      problemId,
    );
  }

  @Post()
  @UseGuards(AuthGuard)
  @ThrottleSubmission()
  async submitInContest(
    @Param('contestId') contestId: string,
    @Param('problemId', ParseIntPipe) problemId: number,
    @Body() dto: CreateSubmissionDto,
    @Req() req: AuthenticatedRequest,
  ) {
    const userId = req.user.id;
    return this.contestSubmissionService.submitInContest(
      contestId,
      problemId,
      userId,
      dto,
    );
  }
}

File: backend-spring/backend-app/src/main/java/com/ulticode/modules/submission/controller/SubmissionController.java
package com.ulticode.modules.submission.controller;

import com.ulticode.websecurity.annotation.RateLimit;
import com.ulticode.common.exception.BusinessException;
import com.ulticode.common.error.BaseErrorCode;
import com.ulticode.common.response.PageResult;
import com.ulticode.common.response.Result;
import com.ulticode.common.auth.CurrentUserProvider;
import com.ulticode.app.api.dto.CreateSubmissionDTO;
import com.ulticode.app.api.dto.LearningProgressDTO;
import com.ulticode.app.api.dto.SubmissionHistoryDTO;
import com.ulticode.app.api.dto.SubmissionQueryDTO;
import com.ulticode.app.api.dto.SubmissionDetailVO;
import com.ulticode.app.api.dto.SubmissionStatusMeta;
import com.ulticode.app.api.dto.SubmissionVO;
import com.ulticode.app.api.service.SubmissionWritePort;
import com.ulticode.modules.submission.projection.SubmissionProjection;
import com.ulticode.modules.submission.service.SubmissionService;
import java.util.List;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.Parameter;
import io.swagger.v3.oas.annotations.media.Content;
import io.swagger.v3.oas.annotations.media.Schema;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.*;

/**
 * REST controller for submission-related operations.
 */
@Tag(name = "Submissions", description = "Code submission endpoints")
@RestController
@RequestMapping("/submissions")
@RequiredArgsConstructor
public class SubmissionController {

    private final SubmissionService submissionService;
    private final SubmissionWritePort submissionWritePort;
    private final SubmissionProjection submissionProjection;
    private final CurrentUserProvider currentUserProvider;

    /**
     * Submit code for a problem.
     * Requires authentication.
     *
     * @param createDTO the submission data
     * @return the created submission
     */
    @Operation(summary = "Submit code", description = "Submit code for a problem")
    @ApiResponse(responseCode = "200", description = "Submission created", content = @Content(schema = @Schema(implementation = SubmissionVO.class)))
    @ApiResponse(responseCode = "400", description = "Validation error")
    @ApiResponse(responseCode = "401", description = "Not authenticated")
    @ApiResponse(responseCode = "404", description = "Problem not found")
    @RateLimit(key = "submission:create", limit = 20, period = 60)
    @PostMapping
    public Result<SubmissionVO> submit(@Valid @RequestBody CreateSubmissionDTO createDTO) {
        String userId = currentUserProvider.getCurrentUserId();
        if (userId == null) {
            throw new BusinessException(BaseErrorCode.UNAUTHORIZED);
        }

        SubmissionVO submission = submissionWritePort.submit(userId, createDTO);
        return Result.success(submission);
    }

    /**
     * Get a submission by ID.
     * Requires authentication.
     *
     * @param id the submission ID
     * @return the submission details
     */
    @Operation(summary = "Get submission by ID",
            description = """
                    Retrieve a specific submission's full detail.

                    **Distribution fields** (`runtimeDistBinsMs` / `memoryDistBinsMb`):
                    Serialized as a JSON array of integers, e.g. `[8, 16, 32, 64, 128, 256, 512]`.
                    Frontend should expect `number[]`, not a JSON string. Legacy callers
                    may still receive a JSON string in transitional windows; the frontend
                    `mapDistributionBins()` helper normalizes both shapes.

                    See `docs/reports/submission-api-test-report-2026-06-10.md` for the
                    full DTO contract.
                    """)
    @ApiResponse(responseCode = "200", description = "Submission retrieved", content = @Content(schema = @Schema(implementation = SubmissionVO.class)))
    @ApiResponse(responseCode = "403", description = "Not authorized to view this submission")
    @ApiResponse(responseCode = "404", description = "Submission not found")
    @GetMapping("/{id}")
    public Result<SubmissionDetailVO> getSubmission(
            @Parameter(description = "Submission ID")
            @PathVariable String id) {
        String userId = currentUserProvider.getCurrentUserId();
        if (userId == null) {
            throw new BusinessException(BaseErrorCode.UNAUTHORIZED);
        }

        SubmissionDetailVO submission = submissionService.findById(id, userId);
        return Result.success(submission);
    }

    /**
     * List user's submissions with pagination.
     * Requires authentication.
     *
     * @param page     the page number (1-based)
     * @param pageSize the number of items per page
     * @param problemId filter by problem ID
     * @return paginated list of submissions
     */
    @Operation(summary = "List user submissions", description = "Get paginated submissions for the authenticated user")
    @ApiResponse(responseCode = "200", description = "Submissions retrieved", content = @Content(schema = @Schema(implementation = PageResult.class)))
    @ApiResponse(responseCode = "401", description = "Not authenticated")
    @GetMapping
    public Result<PageResult<SubmissionVO>> listUserSubmissions(
            @Parameter(description = "Page number (1-based)")
            @RequestParam(required = false) Integer page,
            @Parameter(description = "Number of items per page")
            @RequestParam(required = false) Integer pageSize,
            @Parameter(description = "Filter by problem ID")
            @RequestParam(required = false) Long problemId) {

        String userId = currentUserProvider.getCurrentUserId();
        if (userId == null) {
            throw new BusinessException(BaseErrorCode.UNAUTHORIZED);
        }

        SubmissionQueryDTO query = new SubmissionQueryDTO();
        query.setPage(page);
        query.setPageSize(pageSize);
        query.setProblemId(problemId);

        PageResult<SubmissionVO> result = submissionService.findByUserId(userId, query);
        return Result.success(result);
    }

    /**
     * Get submission calendar (dates with submissions) for the authenticated user.
     * Requires authentication.
     *
     * @param year the year to filter by (defaults to current year)
     * @return list of date strings (YYYY-MM-DD) with submissions
     */
    @Operation(summary = "Get submission calendar", description = "Get dates with submissions for the authenticated user")
    @ApiResponse(responseCode = "200", description = "Calendar retrieved", content = @Content(schema = @Schema(implementation = java.util.List.class)))
    @ApiResponse(responseCode = "401", description = "Not authenticated")
    @GetMapping("/calendar")
    public Result<List<String>> getSubmissionCalendar(
            @Parameter(description = "Year to filter by")
            @RequestParam(required = false) Integer year) {
        String userId = currentUserProvider.getCurrentUserId();
        if (userId == null) {
            throw new BusinessException(BaseErrorCode.UNAUTHORIZED);
        }
        List<String> dates = submissionProjection.aggregateDates(userId, year);
        return Result.success(dates);
    }

    /**
     * Get the best (fastest accepted) submission for a problem.
     * Requires authentication.
     *
     * @param problemId the problem ID
     * @return the best submission, or null if not found
     */
    @Operation(summary = "Get best submission", description = "Get the best accepted submission for a problem")
    @ApiResponse(responseCode = "200", description = "Best submission retrieved", content = @Content(schema = @Schema(implementation = SubmissionVO.class)))
    @ApiResponse(responseCode = "401", description = "Not authenticated")
    @ApiResponse(responseCode = "404", description = "No accepted submission found")
    @GetMapping("/best")
    public Result<SubmissionVO> getBestSubmission(
            @Parameter(description = "Problem ID")
            @RequestParam Long problemId) {

        String userId = currentUserProvider.getCurrentUserId();
        if (userId == null) {
            throw new BusinessException(BaseErrorCode.UNAUTHORIZED);
        }

        SubmissionVO submission = submissionService.findBest(problemId, userId);
        return Result.success(submission);
    }

    /**
     * Get learning progress for the authenticated user.
     * Requires authentication.
     *
     * @return learning progress data including weekly progress, difficulty breakdown, and streaks
     */
    @Operation(summary = "Get learning progress", description = "Get learning progress for the authenticated user")
    @ApiResponse(responseCode = "200", description = "Learning progress retrieved", content = @Content(schema = @Schema(implementation = LearningProgressDTO.class)))
    @ApiResponse(responseCode = "401", description = "Not authenticated")
    @GetMapping("/learning-progress")
    public Result<LearningProgressDTO> getLearningProgress() {
        String userId = currentUserProvider.getCurrentUserId();
        if (userId == null) {
            throw new BusinessException(BaseErrorCode.UNAUTHORIZED);
        }
        LearningProgressDTO progress = submissionProjection.aggregateLearningProgress(userId);
        return Result.success(progress);
    }

    /**
     * Get submission history for the authenticated user.
     * Requires authentication.
     *
     * @return submission history data including monthly stats and language breakdown
     */
    @Operation(summary = "Get submission history", description = "Get submission history for the authenticated user")
    @ApiResponse(responseCode = "200", description = "Submission history retrieved", content = @Content(schema = @Schema(implementation = SubmissionHistoryDTO.class)))
    @ApiResponse(responseCode = "401", description = "Not authenticated")
    @GetMapping("/history")
    public Result<SubmissionHistoryDTO> getSubmissionHistory() {
        String userId = currentUserProvider.getCurrentUserId();
        if (userId == null) {
            throw new BusinessException(BaseErrorCode.UNAUTHORIZED);
        }
        SubmissionHistoryDTO history = submissionProjection.aggregateHistory(userId);
        return Result.success(history);
    }

    /**
     * Get available submission statuses for filtering and display.
     * This endpoint is public (no authentication required).
     *
     * @return list of submission status metadata
     */
    @Operation(summary = "Get submission statuses", description = "Get available submission status options")
    @ApiResponse(responseCode = "200", description = "Statuses retrieved", content = @Content(schema = @Schema(implementation = java.util.List.class)))
    @GetMapping("/statuses")
    public Result<List<SubmissionStatusMeta>> getSubmissionStatuses() {
        List<SubmissionStatusMeta> statuses = submissionProjection.getStatusCatalog();
        return Result.success(statuses);
    }
}
